File: message_pact_js/message.consumer.test.ts
const path = require("path");
const {
  MessageConsumerPact,
  synchronousBodyHandler,
  MatchersV3
} = require("@pact-foundation/pact");


const { eachLike } = MatchersV3;

export enum Wal2JsonChangeKind {
  Update = "update",
  Insert = "insert",
  Delete = "delete"
}

export interface Wal2JsonBase {
  kind: Wal2JsonChangeKind;
  table: string;
  schema: string;
}

export interface Wal2JsonColumns {
  columnnames: string[];
  columnvalues: any[];
  columntypes: string[];
}

interface Wal2JsonOldKeys {
  oldkeys: {
    keytypes: string[];
    keyvalues: any[];
    keynames: string[];
  };
}

export interface Wal2JsonInsert extends Wal2JsonBase, Wal2JsonColumns {}

export interface Wal2JsonUpdate extends Wal2JsonBase, Wal2JsonColumns {}

export interface Wal2JsonDelete extends Wal2JsonBase, Wal2JsonOldKeys {}

export type Wal2JsonChange = Wal2JsonUpdate | Wal2JsonInsert | Wal2JsonDelete;

export interface Wal2JsonEvent {
  xid?: number;
  change: Wal2JsonChange[];
}


// 1 wal2Json change event Handler
const wal2JsonHandler = function (wal2JsonEvent: Wal2JsonEvent) {
  // if (!wal2JsonEvent.xid || !wal2JsonEvent.change) {
  if (!wal2JsonEvent.change) {
    console.log(wal2JsonEvent);
    throw new Error("missing change data");
  }

  // do some other things to the event...
  // like sending it to Kinesis Firehose
  // e.g. wal2JsonEventStreamer.send(wal2JsonEvent)
  // It could get picked up by a lambda transformer to
  // transform into business intelligence logic, or enriched
  // and sent to its final location - uploaded as a csv to a bucket
  return;
};

// 2 Pact Message Consumer
const messagePact = new MessageConsumerPact({
  consumer: "wal2JsonConsumer_js",
  dir: path.resolve(process.cwd(), "pacts"),
  pactfileWriteMode: "update",
  provider: "wal2JsonProvider"
});

describe("receive wal2json event", () => {
  it("accepts a valid wal2json event", () => {
    // 3 Consumer expectations
    return (
      messagePact
        .given("a wal2json replication slot exists")
        .expectsToReceive("a valid wal2json change event")
        .withContent({
          // xid: like("bar"),
          change: eachLike(
            {
              columnnames: ["_id", "id", "name", "price", "version", "type"],
              columntypes: [
                "integer",
                "text",
                "text",
                "numeric",
                "text",
                "text"
              ],
              columnvalues: [1, "9", "Gem Visa", 99.99, "v1", "CREDIT_CARD"],
              kind: "insert",
              schema: "public",
              table: "product"
            }
          )
        })
        // 4 Verify consumers' ability to handle messages
        .verify(synchronousBodyHandler(wal2JsonHandler))
    );
  });
});

File: message_pact_python/message_consumer_test.py
import os
import logging

LOGGER = logging.getLogger('__name__')


import pytest as pytest
from pact import EachLike, Like, MessageConsumer, Provider, Term
from pact.matchers import from_term, get_generated_values

# Actual message handler, doesn't care about SNS at all!
async def receive_product_update(wal2json_event):
    # https://github.com/pact-foundation/pact-python/pull/313
    # This change needs merging, payload isnt reifed before it is sent to consumer
    LOGGER.warning(get_generated_values(wal2json_event))
    wal2json_event = get_generated_values(wal2json_event)
    if not wal2json_event['change']:
    # if not wal2json_event['xid'] or wal2json_event['change']:
        return 'missing fields'
    return "do something"

CONSUMER_NAME = "wal2JsonConsumer_python"
PROVIDER_NAME = "wal2JsonProvider"
PACT_DIR = os.path.join(os.path.dirname(os.path.realpath(__file__)), "pacts")
@pytest.fixture(scope="session")
def consumer():
    return receive_product_update


@pytest.fixture(scope="session")
def pact():
    pact = MessageConsumer(
        CONSUMER_NAME,
    ).has_pact_with(
        Provider(PROVIDER_NAME),
        pact_dir=PACT_DIR,
    )

    yield pact

@pytest.mark.asyncio
async def test_receive_a_product_update(pact, consumer):
    event = {
          "change": EachLike(
            {
              "columnnames": ["_id", "id", "name", "price", "version", "type"],
              "columntypes": [
                "integer",
                "text",
                "text",
                "numeric",
                "text",
                "text"
              ],
              "columnvalues": [1, "9", "Gem Visa", 99.99, "v1", "CREDIT_CARD"],
              "kind": Term(matcher="^(delete|update|insert)$", generate="insert"),
              "schema": "public",
              "table": "product"
            }
          )
        }
    
    (
        pact
        .given("a wal2json replication slot exists")
        .expects_to_receive("a valid wal2json change event")
        .with_content(event)
        .with_metadata({"Content-Type": "application/json"})
    )

    with pact:
        await consumer(event)
        print(PACT_DIR)
